uproot4/exceptions.py

```diff
@@ -0,0 +1,79 @@
+from uproot4._util import damerau_levenshtein
+
+
+class KeyInFileError(KeyError):
+    """
+    Exception raised by attempts to find ROOT objects in ``TDirectories``
+    or ``TBranches`` in :py:class:`~uproot4.behaviors.TBranch.HasBranches`, which
+    both have a Python ``Mapping`` interface (square bracket syntax to extract
+    items).
+
+    This exception descends from Python's ``KeyError``, so it can be used in
+    the normal way by interfaces that expect a missing item in a ``Mapping``
+    to raise ``KeyError``, but it provides more information, depending on
+    availability:
+
+    * ``because``: an explanatory message
+    * ``cycle``: the ROOT cycle number requested, if any
+    * ``keys``: a list or partial list of keys that *are* in the object, in case
+      of misspelling
+    * ``file_path``: a path (or URL) to the file
+    * ``object_path``: a path to the object within the ROOT file.
+    """
+
+    def __init__(
+        self, key, because="", cycle=None, keys=None, file_path=None, object_path=None
+    ):
+        super(KeyInFileError, self).__init__(key)
+        self.key = key
+        self.because = because
+        self.cycle = cycle
+        self.keys = keys
+        self.file_path = file_path
+        self.object_path = object_path
+
+    def __str__(self):
+        if self.because == "":
+            because = ""
+        else:
+            because = " because " + self.because
+
+        with_keys = ""
+        if self.keys is not None:
+            to_show = None
+            distance = []
+            for key in self.keys:
+                distance.append(damerau_levenshtein(self.key, key))
+            sorted_keys = [x for _, x in sorted(zip(distance, self.keys), key=lambda pair: pair[0])]
```

Comment:
Isn't it possible to replace the four lines above with

```python
sorted_keys = sorted(self.keys, key=lambda x: damerau_levenshtein(self.key, x))
```

I haven't explicitly tried it, but it's supposed to work something like that—much simpler (and therefore easier to maintain) than what you have. I might have misled you when I was talking about `cmp` vs `key`; I hadn't noticed that the relevant `key` function here has only one parameter because `self.key` (the bad one that the user tried) is a constant across all `self.keys` (the good alternatives; any one of which is what they meant). Only one argument of `damerau_levenshtein` varies.

I don't see any other issues (now that I'm looking closely).

Oh, I don't know if there are any tests that sample this because it's an exception. I'll take your word for it if, after the above fix, you get the right results in a KeyInFileError. Thanks!

Reply:
You're right. That works.

Reply:
In principle, there could be a test to capture the error and check that it's always the same for some test file, but that sounds a bit extreme to me.

Reply:
A test that checks the string representation of a key error would be extremely brittle. One little change in formatting would cause such a test to fail, and that would be a "false positive" in the sense that nothing actually went wrong, and yet we have to fix tests. A good test suite is as much about what you _don't_ test as what you do test.

Anyway, this looks good. Thanks!